### PR TITLE
Trying to make the stdlib docs more useful, round 1

### PR DESCRIPTION
**Pull request by bethaitman**
Not perfect, but hopefully an improvement.

### CODE REVIEW

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml [hunk 1]
```diff
@@ -57,12 +57,23 @@ infixr 9 .
 const                   : a -> b -> a
 const x _               =  x
 
+-- | A `Functor` is a typeclass for things that can be mapped over (using
+-- its `fmap` function. Examples include `Maybe` and `[]`.
+
+-- For a detailed explanation, see 
+-- http://learnyouahaskell.com/making-our-own-types-and-typeclasses#the-functor-typeclass
+
 class Functor f where
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]: I think you need to remove the blank lines, replacing them with `--`, or the docs won't show up.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml [hunk 2]
```diff
@@ -57,12 +57,23 @@ infixr 9 .
 const                   : a -> b -> a
 const x _               =  x
 
+-- | A `Functor` is a typeclass for things that can be mapped over (using
+-- its `fmap` function. Examples include `Maybe` and `[]`.
+
+-- For a detailed explanation, see 
+-- http://learnyouahaskell.com/making-our-own-types-and-typeclasses#the-functor-typeclass
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]: https: links where possible.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml [hunk 3]
```diff
@@ -57,12 +57,23 @@ infixr 9 .
 const                   : a -> b -> a
 const x _               =  x
 
+-- | A `Functor` is a typeclass for things that can be mapped over (using
+-- its `fmap` function. Examples include `Maybe` and `[]`.
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]: Missing closing bracket

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml [hunk 4]
```diff
@@ -21,97 +21,55 @@ import GHC.CString (fromString)
...
+    -- | Convert a value to a readable `Text` value.
     show      : a   -> Text
 
-    -- | The method 'showList' is provided to allow the programmer to
-    -- give a specialised way of showing lists of values.
-    -- For example, this is used by the predefined 'Show' instance of
-    -- the 'Char' type, where values of type 'String' should be shown
-    -- in double quotes, rather than between square brackets.
+    -- | Allows you to show lists of values.
+    -- For example, this is used by the `Char` type, where values of type
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml]: We don't have Char

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 5]
```diff
@@ -69,61 +66,59 @@ class  Bounded a  where
...
-    --   For example:
+    -- | Returns a list of the `Enum` values with the first value at
+    -- the first `Int` position, the second value at the second `Int`
+    -- position, and further values with the same distance between them,
+    -- with the final value at the final `Int` position.
+    --
+    -- For example:
     --
-    --     * `enumFromThenTo 4 2 -6 :: [Integer] = [4,2,0,-2,-4,-6]`
+    --     * `enumFromThenTo 4 2 -6 :: [Int] = [4,2,0,-2,-4,-6]`
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: Use `:` for type signatures here

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 6]
```diff
@@ -69,61 +66,59 @@ class  Bounded a  where
...
     --     * `enumFromThenTo 6 8 2 :: [Int] = []`
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: Ditto

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 7]
```diff
@@ -69,61 +66,59 @@ class  Bounded a  where
...
-    --     | otherwise = y`
-    --   For example:
+    -- | Returns a list of the `Enum` values with the first value at
+    -- the first `Int` position, the second value at the second `Int`
+    -- position, and further values with the same distance between them.
+    --
+    -- For example:
     --
-    --     * `enumFromThen 4 6 :: [Integer] = [4,6,8,10...]`
+    --     * `enumFromThen 4 6 :: [Int] = [4,6,8,10...]`
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: : for type signatures

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml (hunk 3 above)
Comment (hurryabit) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]:
```suggestion
-- its `fmap` function. Examples include `Optional`, `[]` and `Update`).
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml [hunk 8]
```diff
@@ -57,12 +57,23 @@ infixr 9 .
...
+
+-- For a detailed explanation, see 
+-- http://learnyouahaskell.com/making-our-own-types-and-typeclasses#the-functor-typeclass
+
 class Functor f where
+    -- | `fmap` must take a function from type `a` to type `b`, and a 
+    -- `Functor` of type `a`, and return a `Functor` of type `b`.
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]: I would probably slightly rephrase this as something like:
`fmap` takes a function of type `a -> b` and turns it onto a function of type `f a -> f a`, where `f` is the type which is an instance of `Functor`.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 9]
```diff
@@ -25,40 +25,37 @@ import GHC.Show()
...
 
--- | The `Bounded` class is used to name the upper and lower limits of a
--- type.  `Ord` is not a superclass of `Bounded` since types that are not
--- totally ordered may also have upper and lower bounds.
+-- | Use the `Bounded` class to name the upper and lower limits of a
+-- type.
 --
--- The `Bounded` class may be derived for any enumeration type;
--- `minBound` is the first constructor listed in the `data` declaration
+-- You can derive `Bounded` class for any enumeration type. `minBound`
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]:
```suggestion
-- You can derive an instance of the `Bounded` class for any enumeration type. `minBound`
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 10]
```diff
@@ -25,40 +25,37 @@ import GHC.Show()
...
+-- | Use the `Enum` class to define operations on sequentially ordered
+-- types: that is, types that can be enumerated. `Enum` members have
+-- defined successors and predecessors, which you can get with the `succ`
+-- and `pred` functions.
 --
 -- For any type that is an instance of class `Bounded` as well as `Enum`,
 -- the following should hold:
 --
--- * The calls `succ` `maxBound` and `pred` `minBound` should result in
+-- * Both `succ` `maxBound` and `pred` `minBound` should result in
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]:
```suggestion
-- * Both `succ maxBound` and `pred minBound` should result in
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 11]
```diff
@@ -69,61 +66,59 @@ class  Bounded a  where
...
-    -- | the predecessor of a value.  For numeric types, 'pred' subtracts 1.
+    -- | Returns the predecessor of the given value. For example, for
+    -- numeric types, `pred` subtracts 1.
+    --
+    -- If the type is also an instance of `Bounded`, `pred minBound`
+    -- results in a runtime error.
     pred                : a -> a
-    -- | Convert from an 'Int'.
+    -- | Convert a value from an `Int` to an `Enum` value: ie, returns
+    -- the item at the `Int` position from the `Enum`.
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: I'd say: `toEnum i` returns the item at the `i`th position of (the instance of) `Enum`.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml [hunk 12]
```diff
@@ -69,61 +66,59 @@ class  Bounded a  where
...
-    --   a possible implementation being `enumFrom n = n : enumFrom (succ n)`.
-    --   For example:
+    -- | Return a list of the `Enum` values starting at the `Int`
+    -- position. For example:
     --
-    --     * `enumFrom 4 :: [Integer] = [4,5,6,7,...]`
-    --     * `enumFrom 6 :: [Int] = [6,7,8,9,...,maxBound :: Int]`
+    --     * `enumFrom 6 : [Int] = [6,7,8,9,...,maxBound : Int]`
     enumFrom            : a -> [a]
-    -- | Used in Haskell's translation of `[n,n'..]`
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: I think this information is important. Same below for `enumFromTo`.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 13]
```diff
@@ -33,23 +33,23 @@ infixr 8 ^
...
---
--- **Commutativity of (+)**: `x + y` = `y + x`
---
--- `x + aunit` = `x`
---
--- `negate` **gives the additive inverse**: `x + negate x` = aunit
+-- * (+) must be associative, ie:`(x + y) + z` = `x + (y + z)`
+-- * (+) must be commutative, ie: `x + y` = `y + x`
+-- * `x + aunit` = `x`
+-- * `negate` gives the additive inverse, ie: `x + negate x` = aunit
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]:
```suggestion
-- * `negate` gives the additive inverse, ie: `x + negate x` = `aunit`
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 14]
```diff
@@ -33,23 +33,23 @@ infixr 8 ^
 infixl 7 /
 infixl 7 %
 
--- | Basic additive class.
--- Instances are expected to respect the following laws:
+-- | Use the `Additive` class for types that can be added.
+-- Instances have to follow these rules:
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]: I think "respect the following laws" is a better formulation. After all, they are called "laws" in basic arithmetic. Same for multiplicative below.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 15]
```diff
@@ -33,23 +33,23 @@ infixr 8 ^
...
+-- * (+) must be associative, ie:`(x + y) + z` = `x + (y + z)`
+-- * (+) must be commutative, ie: `x + y` = `y + x`
+-- * `x + aunit` = `x`
+-- * `negate` gives the additive inverse, ie: `x + negate x` = aunit
 --
 class Additive a  where
     {-# MINIMAL (+), aunit, (negate | (-)) #-}
-    (+), (-)            : a -> a -> a
-    -- Unary negation.
+    -- | Add the first two arguments together
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]:
```suggestion
    -- | Add the two arguments together
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 16]
```diff
@@ -58,37 +58,38 @@ class Additive a  where
...
--- **Associativity of (*)**:`(x * y) * z` = `x * (y * z)`
---
--- **Commutativity of (*)**: `x * y` = `y * x`
+-- | Use the `Multiplicative` class for types that can be multiplied.
+-- Instances have to follow these rules:
 --
--- `x * one` = `x`
+-- * (*) is associative, ie:`(x * y) * z` = `x * (y * z)`
+-- * (*) is commutative, ie: `x * y` = `y * x`
+-- * `x * one` = `x`
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]:
```suggestion
-- * `x * munit` = `x`
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 17]
```diff
@@ -58,37 +58,38 @@ class Additive a  where
...
     (*) : a -> a -> a
+    -- | The multiplicative identity for the type. For example, for numbers, this is 1.
     munit : a
     -- | `x ^ n` raises `x` to the power of `n`.
     (^): a -> Int -> a
 
--- | Basic numeric class.
--- Instances are usually expected to respect the following law
--- (in addition to laws from being additive and multiplicative)
+-- | `Number` is a class for basic numerical types.
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]:
```suggestion
-- | `Number` is a class for numerical types.
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 18]
```diff
@@ -58,37 +58,38 @@ class Additive a  where
...
+-- As well as the rules for `Additive` and `Multiplicative`, instances
+-- also have to follow these rules:
 --
--- **Distributivity of (*) with respect to (+)**:
+-- * (*) is distributive with respect to (+), ie:
 --     `a * (b + c)` = `(a * b) + (a * c)` and `(b + c) * a` = `(b * a) + (c * a)`
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]: Since `*` has to be commutative, we could scrap the second equation. I'm not sure this is a great idea though. @neil-da?

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 19]
```diff
@@ -58,37 +58,38 @@ class Additive a  where
...
 class (Additive a, Multiplicative a) => Number a where
   -- no methods
 
+-- | The `Signed` is for the sign of a number.
 class Signed a where
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]: @neil-da Should we add a super class constraint `Number a` to this? What else does have a sign?

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 20]
```diff
@@ -33,23 +33,23 @@ infixr 8 ^
...
 --
 class Additive a  where
     {-# MINIMAL (+), aunit, (negate | (-)) #-}
-    (+), (-)            : a -> a -> a
-    -- Unary negation.
+    -- | Add the first two arguments together
+    (+)                 : a -> a -> a
+    -- | Subtract the second argument from the first argument
+    (-)                 : a -> a -> a
+    -- | Negate the argument: `x + negate x` = the additive identity for the type
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]:
```suggestion
    -- | Negate the argument: `x + negate x` = `aunit`
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 21]
```diff
@@ -33,23 +33,23 @@ infixr 8 ^
...
+-- * `x + aunit` = `x`
+-- * `negate` gives the additive inverse, ie: `x + negate x` = aunit
 --
 class Additive a  where
     {-# MINIMAL (+), aunit, (negate | (-)) #-}
-    (+), (-)            : a -> a -> a
-    -- Unary negation.
+    -- | Add the first two arguments together
+    (+)                 : a -> a -> a
+    -- | Subtract the second argument from the first argument
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]:
```suggestion
    -- | Subtract the second argument from the first argument, ie. `x - y` = `x + negate y`
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml [hunk 22]
```diff
@@ -132,11 +133,12 @@ instance Signed Decimal where
     signum x = if x == 0.0 then 0.0 else if x <= 0.0 then (-1.0) else 1.0
     abs x = if x <= 0.0 then negate x else x
 
+-- | The `Fractional` class.
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]: We should expand to say this is for types that can be divided. `recip x` does not need to be defined for all `x : a`, for instance it is not defined for `0.0 : Decimal`. There are also some laws:
- when `recip x` is defined, it need to be the inverse of `x` wrt `*`, ie. `x * recip` = `munit` (in fact, our `Decimal` types only obeys this law only up to some precision)
- when `recip y` is defined, then `x / y` = `x * recip y`

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml [hunk 23]
```diff
@@ -21,97 +21,55 @@ import GHC.CString (fromString)
 import GHC.Tuple ()  -- We're using tuples but they are builtin syntax.
 import GHC.Types
 
--- | The `shows` functions return a function that prepends the
--- output 'String' to an existing 'String'.  This allows constant-time
+-- | `showS` is a function that prepends the
+-- output `Text` to an existing `Text`.  This allows constant-time
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml]: This is pretty tricky to document since we can only document the _intent_ of a function of type `ShowS`. Maybe something like:
A function of type `ShowS` _should_ represent some text and applying it so some argument _should_ prepend the argument to the represented text.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml [hunk 24]
```diff
@@ -21,97 +21,55 @@ import GHC.CString (fromString)
...
-    -- That is, `Text.Read.readsPrec` parses the string produced by
-    -- `showsPrec`, and delivers the value that `showsPrec` started with.
+    -- | Convert a value to a readable `Text` value. Unlike `show`,
+    -- showsPrec` should satisfy the rule `showsPrec d x r ++ s  ==
+    -- showsPrec d x (r ++ s)`
 
     showsPrec : Int    -- ^ the operator precedence of the enclosing
                         -- context (a number from `0` to `11`).
                         -- Function application has precedence `10`.
               -> a      -- ^ the value to be converted to a 'String'
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml]:
```suggestion
              -> a      -- ^ the value to be converted to a `Text`
```

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml (hunk 4 above)
Comment (hurryabit) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Show.daml]: @neil-da Shall we just drop `showList` from the class?

File: daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/LF.daml [hunk 25]
```diff
@@ -128,16 +128,17 @@ instance Action Update where
...
-    -- | Obtain the current time.
+    -- | Get the current time.
     getTime : m Time
 
 instance HasTime Update where
     getTime = primitive @"UGetTime"
 instance HasTime Scenario where
     getTime = primitive @"SGetTime"
 
+-- | The `CanAbort` class is for Updates that can be aborted.
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/LF.daml]:
```suggestion
-- | The `CanAbort` class is for `Action`s that can be aborted.
```

File: daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/LF.daml [hunk 26]
```diff
@@ -128,16 +128,17 @@ instance Action Update where
 instance ActionFail Update where
     fail = primitive @"UAbort"
 
--- | Places where the time is available - basically `Scenario` and `Update`.
+-- | The `HasTime` class is for where the time is available: `Scenario` and `Update`.
 class HasTime m where
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/LF.daml]: @neil-da Should we add a `Action m` super class constraint? Then the documentation could be:
The `HasTime` class is for `Action`s that support a notion of current time, like `Update` and `Scenario`.

File: daml-foundations/daml-ghc/src/DA/Daml/GHC/Damldoc/Render/Markdown.hs [hunk 27]
```diff
@@ -33,12 +33,14 @@ renderSimpleMD ModuleDoc{..} = T.unlines $
   , if null md_classes
     then []
     else [ "## Typeclasses"
+         , ""
+         , "Typeclasses in DAML are an interface that defines some behavior."
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/src/DA/Daml/GHC/Damldoc/Render/Markdown.hs]: Isn't this mixture of singular and plural odd?

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml (hunk 1 above)
Comment (bethaitman) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]: Ok, good to know! I meant to generate the docs to test this and forgot - will do today.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml (hunk 12 above)
Comment (bethaitman) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: Maybe it's just the phrasing I object to. I might rephrase to say that this is what allows you to do `[n,n'..]`.

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml (hunk 5 above)
Comment (bethaitman) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Enum.daml]: It would be so good if we could actually test these for correctness

File: daml-foundations/daml-ghc/src/DA/Daml/GHC/Damldoc/Render/Markdown.hs (hunk 27 above)
Comment (bethaitman) [reply daml-foundations/daml-ghc/src/DA/Daml/GHC/Damldoc/Render/Markdown.hs]: yeah, you're right

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml (hunk 2 above)
Comment (bethaitman) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Base.daml]: unfortunately the site does actually use http, not https

File: daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml (hunk 22 above)
Comment (bethaitman) [reply daml-foundations/daml-ghc/daml-prim-src/GHC/Num.daml]: Added!

**Comment by bethaitman** [comment]
Thanks for all the comments, both! I've addressed them all and fixed a few formatting bugs as well, now adding the `automerge` label.
